File: testSuite/test-inactive_luminosities.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib $ENV{'GALACTICUS_EXEC_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;

# Test inactive luminosity calculations give results identical to active luminosity calculations.
# Andrew Benson (14-May-2019)

# Specify types.
my @types = ( 'active', 'inactive' );

# Run the models.
system("mkdir -p outputs/inactiveLuminosities");
foreach my $type ( @types ) {
    system("cd ..; ./Galacticus.exe testSuite/parameters/".$type."Luminosities.xml");
    die("FAILED: model '".$type."' failed to run")
	unless ( $? == 0 );
}

# Extraxt datasets and compare.
my $models;
foreach my $type ( @types ) {
    $models->{$type}->{'file'   } = new PDL::IO::HDF5("outputs/inactiveLuminosities/".$type."Luminosities.hdf5");
    $models->{$type}->{'outputs'} = $models->{$type}->{'file'}->group('Outputs');
}
my $success = 1;
print "Status\tDataset\tFractional error\n";
for(my $i=1;$i<=4;++$i) {
    foreach my $type ( @types ) {
	$models->{$type}->{'nodes'.$i} = $models->{$type}->{'outputs'}->group('Output'.$i)->group('nodeData');
    }
    my @datasetNames = $models->{'active'}->{'nodes'.$i}->datasets();
    foreach my $datasetName ( @datasetNames ) {
	next
	    unless
	    (
	     $datasetName =~ m/^(disk|spheroid)MassStellar$/
	     ||
	     $datasetName =~ m/^(disk|spheroid)LuminositiesStellar/
	    );
	my $property;
	foreach my $type ( @types ) {
	    $property->{$type} = $models->{$type}->{'nodes'.$i}->dataset($datasetName)->get();
	}
	my $errorFractional = abs($property->{'inactive'}-$property->{'active'})/$property->{'active'};
	my $status = "SUCCESS";
	if ( $errorFractional->((0)) > 1.5e-3 ) {
	    $status  = "FAILED";
	    $success = 0;
	}
	print $status."\t".$datasetName."\t".$errorFractional->((0))."\n";
    }
}
my $status = $success ? "SUCCESS" : "FAILED: some datasets differed";
print "\n".$status."\n";
exit;

File: testSuite/test-inactive_luminosities.py
#!/usr/bin/env python3
import subprocess
import sys
import re
import h5py
import numpy as np

# Test inactive luminosity calculations give results identical to active luminosity calculations.
# Andrew Benson (ported to Python)

# Specify types.
types = ("active", "inactive")

# Run the models.
subprocess.run("mkdir -p outputs/inactiveLuminosities", shell=True)
for ltype in types:
    status = subprocess.run(f"cd ..; ./Galacticus.exe testSuite/parameters/{ltype}Luminosities.xml", shell=True)
    if status.returncode != 0:
        print(f"FAILED: model '{ltype}' failed to run")
        sys.exit(0)

# Extract datasets and compare.
models = {}
for ltype in types:
    models[ltype] = h5py.File(f"outputs/inactiveLuminosities/{ltype}Luminosities.hdf5", "r")

success = True
print("Status\tDataset\tFractional error")
try:
    for i in range(1, 5):
        nodes = {ltype: models[ltype][f"Outputs/Output{i}/nodeData"] for ltype in types}
        datasetNames = list(nodes["active"].keys())
        for datasetName in datasetNames:
            if not (re.match(r'^(disk|spheroid)MassStellar$', datasetName) or
                    re.match(r'^(disk|spheroid)LuminositiesStellar', datasetName)):
                continue
            prop = {}
            for ltype in types:
                prop[ltype] = nodes[ltype][datasetName][:]
            errorFractional = np.abs(prop["inactive"] - prop["active"]) / prop["active"]
            status_str = "SUCCESS"
            if errorFractional[0] > 1.5e-3:
                status_str = "FAILED"
                success    = False
            print(f"{status_str}\t{datasetName}\t{errorFractional[0]}")
finally:
    for ltype in types:
        models[ltype].close()

status_str = "SUCCESS" if success else "FAILED: some datasets differed"
print(f"\n{status_str}")
